### PR TITLE
#3510: Unify CS loading detection; avoid side effects

### PR DESCRIPTION
**Pull request by fregante**
## Context

This PR is part of an effort to ensure loading is done correctly and to avoid inexplicable bugs caused by delayed or duplicate injections:

- comment: https://github.com/pixiebrix/pixiebrix-extension/issues/3510#issuecomment-1140509405
- comment: https://github.com/pixiebrix/pixiebrix-extension/pull/3718/files#r897457833
- related: https://github.com/pixiebrix/pixiebrix-extension/pull/3891 

## Why this PR

The deduplication logic had a few problems:

- it assumed code "before other imports" was run before them
- it imported files (and their side effects) even if duplication was later detected
- it used multiple/overlapping ways to declare state (symbols, attributes) — parts of the code used symbol, others attributes
- it had declarations of constants in several different files
- the attribute checker was setting the attribute first and then checking whether it existed (ie always true)

## What does this PR do?

- ~~drops symbol-based detection~~ _makes detection consistent and without duplicats_
- moves all constants to ~~@/common~~ __@/contentScript/ready__
- fixes, documents, and logs the deduplication logic and actions taken at every step
- extracts the contentScript core to a separate file that is only loaded after the deduplication check (instant) is run

The last one especially:

- makes the detection faster
- avoids duplicate registration of global listeners that happen as side effects of some imports (this might have been causing weird bugs we never understood)
- correctly times the CS loading time

## Generated files

- contentScript.js: 1100KB -> 37KB
- contentScriptCore.js: +360KB (I assume this also somehow triggers some deduplication and it's able to load other chunks)

## Fixes

This PR fixes two similar behaviors that happen when the editor is open 🎉 

- https://github.com/pixiebrix/pixiebrix-extension/issues/3132#issuecomment-1094554105
- https://github.com/pixiebrix/pixiebrix-extension/issues/3510#issuecomment-1200262862

### CODE REVIEW

**Comment by fregante** [comment]
In the meantime I found out that the DOM-based check added by https://github.com/pixiebrix/pixiebrix-extension/pull/3718 did not work and in reality we cannot use a DOM check.

We _do_ want a new content script if a context is invalidated. I think we can/should to back to the Symbol-based check, but this PR highlighted some more issues so I need to work on it further… potentially solving https://github.com/pixiebrix/pixiebrix-extension/issues/3132

File: src/pageEditor/context.ts [hunk 1]
```diff
@@ -153,20 +154,19 @@ export function useDevConnection(): Context {
...
+    if (!isActive()) {
+      return;
+    }
+
+    setTabState({
+      ...initialFrameState,
+      navSequence: uuidv4(),
+      error:
+        "The connection to the PixieBrix browser extension was lost. Reload the Page Editor.",
+    });
```
Comment (fregante) [inline src/pageEditor/context.ts:169]: Does this change look right? It works

File: src/pageEditor/context.ts [hunk 2]
```diff
@@ -153,20 +154,19 @@ export function useDevConnection(): Context {
...
-          navSequence: uuidv4(),
-          error:
-            "The connection to the PixieBrix browser extension was lost. Reload the Page Editor.",
-        });
-      }
-    }, 500);
-    return () => {
-      clearInterval(interval);
-    };
+  useAsyncEffect(async (isActive) => {
```
Comment (BALEHOK) [inline src/pageEditor/context.ts:157]: I think `isMounted` is a more common name for the variable, [see example](https://github.com/pixiebrix/pixiebrix-extension/blob/318c9930c46d69448676c12dc3eaf90a4d72e473/src/contrib/google/sheets/FileWidget.tsx#L63)

It reflects if the React component is still in the React tree and is present in the DOM.
`iaActive` is something that we often use in relation with an active tab, or window, or an item selected in the UI.

File: src/background/util.ts [hunk 3]
```diff
@@ -26,6 +26,9 @@ import pTimeout from "p-timeout";
 import type { Target } from "@/types";
 import { getTabsWithAccess } from "./activeTab";
 
+// eslint-disable-next-line import/no-restricted-paths -- TODO: Migrate to webext-messenger?
+import { getTargetState } from "@/contentScript/ready";
```
Comment (twschiller) [inline src/background/util.ts:30]: 👍

File: src/contentScript/contentScriptCore.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  addListenerForUpdateSelectedElement();
+  initTelemetry();
+  initToaster();
+
+  const sender = await whoAmI();
+
+  updateTabInfo({ tabId: sender.tab.id, frameId: sender.frameId });
+
+  await handleNavigate();
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/contentScript/contentScriptCore.ts:68]: handleNavigate is the method that re-activates extensions on the page whenever there is a navigation event

The call in init() is for the "initial" navigation -- the initial page load

- Installs the page script (if not already installed): https://github.com/pixiebrix/pixiebrix-extension/blob/a23280c10a8032e0b487f53a917e4469706781cb/src/contentScript/lifecycle.ts#L378-L378
- Runs all the extension points

**Review by twschiller — APPROVED** [review]

File: src/contentScript/contentScriptCore.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (fregante) [reply src/contentScript/contentScriptCore.ts:68]: Thanks!

I'll merge this as is for now to minimize the chance of breakage. We can review the readiness speed later.

**Pull request merged**